server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -4181,10 +4183,25 @@ public void waitForSegmentGeneration(long segmentGeneration, ActionListener<Long
         waitForPrimaryTermAndGeneration(getOperationPrimaryTerm(), segmentGeneration, listener);
     }
 
+    private final SubscribableListener<Void> waitForPrimaryTermAndGenerationListeners = new SubscribableListener<Void>();
+
+    private void checkAndCallWaitForPrimaryTermAndGenerationListeners() {
+        if (state == IndexShardState.STARTED || state == IndexShardState.CLOSED) {
+            waitForPrimaryTermAndGenerationListeners.onResponse(null);
+        }
+    }
+
     /**
      * Registers a listener for an event when the shard advances to the provided primary term and segment generation
      */
     public void waitForPrimaryTermAndGeneration(long primaryTerm, long segmentGeneration, ActionListener<Long> listener) {
-        getEngine().addPrimaryTermAndGenerationListener(primaryTerm, segmentGeneration, listener);
+        waitForPrimaryTermAndGenerationListeners.addListener(
+            ActionListener.wrap(
+                (ignored) -> getEngine().addPrimaryTermAndGenerationListener(primaryTerm, segmentGeneration, listener),
+                listener::onFailure
+            )
+        );
```

Comment:
Just thinking out loud: The listener will be executed either on this calling thread (same as the existing behaviour) or any other thread that calls `checkAndCallWaitForPrimaryTermAndGenerationListeners` in the future when it first completes (new behaviour). I think the new behaviour is OK since `addPrimaryTermAndGenerationListener` is pretty lightweight and can be executed inline, i.e. without forking to another threadpool.